### ISSUE TITLE
SoftmaxTransform Operator

### ISSUE THREAD

**Issue by Kublai-Jing** [dmlc/mxnet#472]
For some reason, I wanna use Softmax as a transformation and use my own loss function for backprop. However, I cannot find a place to change this setting and currently, it seems that even if I explicitly send grad to executor.backward([grad]). It still uses the Softmax default gradient...

I've seen that there is this "need_top_grad" in Naive_ops, but where can I find such thing for Softmax?

Thanks

**Comment by piiswrong** [comment]
Softmax (recently renamed to SoftmaxOutput) is a loss layer. It's not intended to be used as internal layers. There was a discussion on this: https://github.com/dmlc/mxnet/issues/426

However there is an easy patch for your need you can do. Cudnn provide softmax implementation for internal layers. You can modify the cudnn_activation-inl.h file under src/operator and add a new activation with cudnn API.

If you figured out a solution, your are welcome to contribute to mxnet by submitting PR.

**Comment by tqchen** [comment]
This is discussed to be added as SoftmaxTransform Operator. cc @antinucleon  Please let us know if you are interested in contributing

**Comment by antinucleon** [comment]
See if I have time tomorrow

**Comment by Kublai-Jing** [comment]
I will ry to submit a PR today

**Comment by Kublai-Jing** [comment]
Should that be a stand-alone operator or should it be included in the activation operator ?

**Comment by tqchen** [comment]
I think we can do it in a standalone one, as this is quite different from elementwise activations

**Comment by piiswrong** [comment]
Then we need both softmaxtransform and cudnn softmax transform

**Comment by tqchen** [comment]
These two can be done in one ops, like currently Convolution

**Comment by antinucleon** [comment]
Agree to @piiswrong

**Comment by tqchen** [comment]
Let us see if there is any updates on this

**Comment by piiswrong** [comment]
Softmax transform is now supported via layer SoftmaxActivation.
